aws-codeguruprofiler-profilinggroup/src/main/java/software/amazon/codeguruprofiler/profilinggroup/CreateHandler.java

```diff
@@ -17,42 +21,90 @@
+
+        Optional<List<String>> principals = principalsForAgentPermissionsFrom(model);
+        if (principals.isPresent()) {
+            putAgentPermissions(proxy, pgName, principals.get());
+            logger.log(format("%s [%s] for accountId [%s] has been successfully updated with agent permissions!",
+                    ResourceModel.TYPE_NAME, pgName, awsAccountId));
+        }
+
+        return ProgressEvent.defaultSuccessHandler(model);
+    }
 
-            logger.log(String.format("%s [%s] for accountId [%s] has been successfully created!", ResourceModel.TYPE_NAME, model.getProfilingGroupName(), awsAccountId));
+    private void putAgentPermissions(final AmazonWebServicesClientProxy proxy, final String pgName, final List<String> principals) {
+        PutPermissionRequest putPermissionRequest = PutPermissionRequest.builder()
+                .profilingGroupName(pgName)
+                .actionGroup(ActionGroup.AGENT_PERMISSIONS)
+                .principals(principals)
+                .build();
+        try {
+            proxy.injectCredentialsAndInvokeV2(putPermissionRequest, profilerClient::putPermission);
+        } catch (CodeGuruProfilerException e) {
+            try {
+                DeleteProfilingGroupRequest deletePgRequest = DeleteProfilingGroupRequest.builder().profilingGroupName(pgName).build();
+                proxy.injectCredentialsAndInvokeV2(deletePgRequest, profilerClient::deleteProfilingGroup);
+            } catch (Throwable deleteException) {
+                e.addSuppressed(deleteException);
+            } finally {
+                wrapException(e);
+            }
+        }
+    }
 
-            return ProgressEvent.defaultSuccessHandler(model);
-        } catch (ConflictException e) {
+    private static void wrapException(final CodeGuruProfilerException e) {
+        if (e instanceof ConflictException) {
             throw new CfnAlreadyExistsException(e);
-        } catch (InternalServerException e) {
+        }
+        if (e instanceof InternalServerException) {
             throw new CfnServiceInternalErrorException(e);
-        } catch (ServiceQuotaExceededException e) {
+        }
+        if (e instanceof ServiceQuotaExceededException) {
             throw new CfnServiceLimitExceededException(e);
-        } catch (ThrottlingException e) {
+        }
+        if (e instanceof ThrottlingException) {
             throw new CfnThrottlingException(e);
-        } catch (ValidationException e) {
+        }
+        if (e instanceof ValidationException) {
             throw new CfnInvalidRequestException(ResourceModel.TYPE_NAME + e.getMessage(), e);
         }
+        throw e;
+    }
+
+    private static Optional<List<String>> principalsForAgentPermissionsFrom(final ResourceModel model) {
+        if (model.getAgentPermissions() == null) {
+            return Optional.empty();
+        }
+        return Optional.ofNullable(model.getAgentPermissions().getPrincipals());
```

Comment:
I do not think you would need Optional and Nullable here.

As what I tested (with the old schema you had in revision 1), if you have AgentPermissions field defined, you cannot have null for principals/ empty list for principals. That would caused validation error and I think it makes sense to enforce user to give non-null/ non-empty list for principals in the schema so we do not need to deal with all these here. We can simply check if AgentPermission is null or not to distinguish whether we need to attach policy to the PG.

Reply:
I agree with the removing the ```Nullable``` as also Ivo pointed out in a comment, mostly for readability of the code too.

But I would keep the Optional. Even though the schema does some validation, I would not like to rely in code only on validations from the schema. We can consider in the future that the code would be called without the schema and the code should be prepared to not throw a NPE when using the objects.